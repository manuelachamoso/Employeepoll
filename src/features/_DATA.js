let users = {
    sarahedo: {
      id: "sarahedo",
      password: "password123",
      name: "Sarah Edo",
      avatarURL:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQGAbkVgyQbwYNrSwq0FqWDWCJcOQVOf23DvHbOR-G105CfKl0I0Uo6PhBRTHRlianbvHE&usqp=CAU",
      answers: {
        "8xf0y6ziyjabvozdd253nd": "optionOne",
        "6ni6ok3ym7mf1p33lnez": "optionOne",
        am8ehyc8byjqgar0jgpub9: "optionTwo",
        loxhs1bqm25b708cmbf3g: "optionTwo",
      },
      questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
    },
    tylermcginnis: {
      id: "tylermcginnis",
      password: "abc321",
      name: "Tyler McGinnis",
      avatarURL:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSTm-KJMECWJvjbROgmX9cEk8JNFy9lrYNrBP1FV7oZPw&s",
      answers: {
        vthrdm985a262al8qx3do: "optionOne",
        xj352vofupe1dqz9emx13r: "optionTwo",
      },
      questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
    },
    mtsamis: {
      id: "mtsamis",
      password: "xyz123",
      name: "Mike Tsamis",
      avatarURL:
        "https://media.istockphoto.com/id/1227618813/es/vector/human-face-avatar-icon-perfil-para-la-red-social-hombre-ilustraci%C3%B3n-vectorial.jpg?s=612x612&w=0&k=20&c=l-Nb1sWxlo4feSppXnBVCSaZpYkFI7v147ce0Ym4lTQ=",
      answers: {
        xj352vofupe1dqz9emx13r: "optionOne",
        vthrdm985a262al8qx3do: "optionTwo",
        "6ni6ok3ym7mf1p33lnez": "optionOne",
      },
      questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
    },
    zoshikanlu: {
      id: "zoshikanlu",
      password: "pass246",
      name: "Zenobia Oshikanlu",
      avatarURL:
        "https://media.istockphoto.com/id/1227618791/es/vector/human-face-avatar-icon-perfil-para-la-red-social-hombre-ilustraci%C3%B3n-vectorial.jpg?s=170667a&w=0&k=20&c=S36eD7dDlXusPKOULAePQClmmLEQlvPdgo16EhP-slw=",
      answers: {
        xj352vofupe1dqz9emx13r: "optionOne",
      },
      questions: [],
    },
  };
  
  let questions = {
    "8xf0y6ziyjabvozdd253nd": {
      id: '8xf0y6ziyjabvozdd253nd',
      author: 'sarahedo',
      timestamp: 1467166872634,
      optionOne: {
        votes: ['sarahedo'],
        text: 'Build our new application with Javascript',
      },
      optionTwo: {
        votes: [],
        text: 'Build our new application with Typescript'
      }
    },
    "6ni6ok3ym7mf1p33lnez": {
      id: '6ni6ok3ym7mf1p33lnez',
      author: 'mtsamis',
      timestamp: 1468479767190,
      optionOne: {
        votes: [],
        text: 'hire more frontend developers',
      },
      optionTwo: {
        votes: ['mtsamis', 'sarahedo'],
        text: 'hire more backend developers'
      }
    },
    "am8ehyc8byjqgar0jgpub9": {
      id: 'am8ehyc8byjqgar0jgpub9',
      author: 'sarahedo',
      timestamp: 1488579767190,
      optionOne: {
        votes: [],
        text: 'conduct a release retrospective 1 week after a release',
      },
      optionTwo: {
        votes: ['sarahedo'],
        text: 'conduct release retrospectives quarterly'
      }
    },
    "loxhs1bqm25b708cmbf3g": {
      id: 'loxhs1bqm25b708cmbf3g',
      author: 'tylermcginnis',
      timestamp: 1482579767190,
      optionOne: {
        votes: [],
        text: 'have code reviews conducted by peers',
      },
      optionTwo: {
        votes: ['sarahedo'],
        text: 'have code reviews conducted by managers'
      }
    },
    "vthrdm985a262al8qx3do": {
      id: 'vthrdm985a262al8qx3do',
      author: 'tylermcginnis',
      timestamp: 1489579767190,
      optionOne: {
        votes: ['tylermcginnis'],
        text: 'take a course on ReactJS',
      },
      optionTwo: {
        votes: ['mtsamis'],
        text: 'take a course on unit testing with Jest'
      }
    },
    "xj352vofupe1dqz9emx13r": {
      id: 'xj352vofupe1dqz9emx13r',
      author: 'mtsamis',
      timestamp: 1493579767190,
      optionOne: {
        votes: ['mtsamis', 'zoshikanlu'],
        text: 'deploy to production once every two weeks',
      },
      optionTwo: {
        votes: ['tylermcginnis'],
        text: 'deploy to production once every month'
      }
    },
  }
  
  function generateUID () {
    return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
  }
  
  export function _getUsers () {
    return new Promise((resolve) => {
      setTimeout(() => resolve({...users}), 1000)
    })
  }
  
  export function _getQuestions () {
    return new Promise((resolve) => {
      setTimeout(() => resolve({...questions}), 1000)
    })
  }
  
  function formatQuestion ({ optionOneText, optionTwoText, author }) {
    return {
      id: generateUID(),
      timestamp: Date.now(),
      author,
      optionOne: {
        votes: [],
        text: optionOneText,
      },
      optionTwo: {
        votes: [],
        text: optionTwoText,
      }
    }
  }
  
  export function _saveQuestion (question) {
    return new Promise((resolve, reject) => {
      if (!question.optionOneText || !question.optionTwoText || !question.author) {
        reject("Please provide optionOneText, optionTwoText, and author");
      }
  
      const formattedQuestion = formatQuestion(question)
      setTimeout(() => {
        questions = {
          ...questions,
          [formattedQuestion.id]: formattedQuestion
        }
  
        resolve(formattedQuestion)
      }, 1000)
    })
  }
  
  export function _saveQuestionAnswer ({ authedUser, qid, answer }) {
    return new Promise((resolve, reject) => {
      if (!authedUser || !qid || !answer) {
        reject("Please provide authedUser, qid, and answer");
      }
  
      setTimeout(() => {
        users = {
          ...users,
          [authedUser]: {
            ...users[authedUser],
            answers: {
              ...users[authedUser].answers,
              [qid]: answer
            }
          }
        }
  
        questions = {
          ...questions,
          [qid]: {
            ...questions[qid],
            [answer]: {
              ...questions[qid][answer],
              votes: questions[qid][answer].votes.concat([authedUser])
            }
          }
        }
  
        resolve(true)
      }, 500)
    })
  }