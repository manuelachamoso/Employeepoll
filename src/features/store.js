import { configureStore } from "@reduxjs/toolkit";
import currentUserReducer from "../reducers/userSlice";
import currentAnsweredReducer from "../reducers/answeredSlice";
import allQuestionsReducer from "../reducers/questionsSlice";

export const store = configureStore({
  reducer: {
    currentUser: currentUserReducer,
    currentAnswered: currentAnsweredReducer,
    
    questions: allQuestionsReducer,
  },
});



