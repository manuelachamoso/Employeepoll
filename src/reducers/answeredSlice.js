import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  isAnswered: false,
};

const answeredSlice = createSlice({
  name: "currentAnswer",
  initialState,
  reducers: {
    toggleAnswered: (state, action) => {
      state.isAnswered = !state.isAnswered;
    },
  },
});

export const { toggleAnswered } = answeredSlice.actions;

export default answeredSlice.reducer;
