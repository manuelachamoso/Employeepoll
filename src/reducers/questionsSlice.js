import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { _getQuestions, _saveQuestionAnswer, _saveQuestion } from "../features/_DATA";

const initialState = {
  isLoadingQues: false,
  isLoadingAns: false,
  isLoadingQuestions: false,
  questions: [],
};

export const getQuestions = createAsyncThunk(
  "questions/getQuestions",
  async () => {
    let res = await _getQuestions().then((value) => {
      return value;
    });

    return res;
  }
);

export const saveQuestionAnswer = createAsyncThunk(
  "question/updateQuestions",
  async (data) => {
    let res = await _saveQuestionAnswer(data).then((value) => {
      return value;
    });

    return res;
  }
);

export const saveQuestion = createAsyncThunk(
  "question/addQuestion",
  async (question) => {
    let res = await _saveQuestion(question).then((value) => {
      return value;
    });

    return res;
  }
);

const allQuestionsSlice = createSlice({
  name: "allQuestions",
  initialState,
  reducers: {},
  extraReducers: {
    [getQuestions.pending]: (state) => {
      state.isLoadingQuestions = true;
    },
    [getQuestions.fulfilled]: (state, action) => {
      state.isLoadingQuestions = false;
      const quesKeys = Object.keys(action.payload);

      let allQuestions = [];
      quesKeys.map((quesKey) => {
        return allQuestions.push(action.payload[quesKey]);
      });

      state.questions = allQuestions;
    },
    [getQuestions.rejected]: (state, action) => {
      state.isLoadingQuestions = false;
    },

    [saveQuestionAnswer.pending]: (state) => {
      state.isLoadingAns = true;
    },
    [saveQuestionAnswer.fulfilled]: (state, action) => {
      state.isLoadingAns = false;
    },
    [saveQuestionAnswer.rejected]: (state, action) => {
      state.isLoadingAns = false;
    },

    [saveQuestion.pending]: (state) => {
      state.isLoadingQues = true;
    },
    [saveQuestion.fulfilled]: (state, action) => {
      state.isLoadingQues = false;
    },
    [saveQuestion.rejected]: (state, action) => {
      state.isLoadingQues = false;
    },
  },
});

export default allQuestionsSlice.reducer;
