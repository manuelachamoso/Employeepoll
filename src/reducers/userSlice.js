import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { _getUsers } from "../features/_DATA";

const initialState = {
  currentUser: "",
  isLoadingUser: false,
  isLoadingUsers: false,
  users: [],
  currUserQuestions: [],
};

export const getUsers = createAsyncThunk("users/getUsers", async () => {
  let res = await _getUsers().then((value) => {
    return value;
  });

  return res;
});

export const saveUser = createAsyncThunk(
  "user/saveUser",
  async (user) => {
    let res = await _getUsers().then((value) => {
      return value;
    });

    let response = { newuser: res[user.id], users: res };
    return response;
  }
);
const userSlice = createSlice({
  name: "currentUser",
  initialState,
  reducers: {
    setUser: (state, { payload }) => {
      state.currentUser = payload;
    },
    getCurrentUser: (state, { payload }) => {
      state.currentUser = payload;
    },
    setUsers: (state, { payload }) => {
      state.users = payload;
    },
  },
  extraReducers: {
    [getUsers.pending]: (state) => {
      state.isLoadingUsers = true;
    },
    [getUsers.fulfilled]: (state, action) => {
      state.isLoadingUsers = false;

      state.users = action.payload;
    },
    [getUsers.rejected]: (state, action) => {
      state.isLoadingUsers = false;
    },

    [saveUser.pending]: (state) => {
      state.isLoadingUser = true;
    },
    [saveUser.fulfilled]: (state, action) => {
      state.isLoadingUser = false;
    },
    [saveUser.rejected]: (state, action) => {
      state.isLoadingUser = false;
    },
  },
});

export const { setUser, setUsers } = userSlice.actions;

export default userSlice.reducer;

