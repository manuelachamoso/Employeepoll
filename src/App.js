import Navbar from "./components/Navbar";
import { Routes, Route } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import Home from "./components/Home";
import Addpoll from "./components/Addpoll";
import Leaderboard from "./components/Leaderboard";
import Login from "./components/Login";
import Error from "./components/Error";
import Question from "./components/Question";
import { getQuestions } from "./reducers/questionsSlice";
import { getUsers } from "./reducers/userSlice";
import { useEffect } from "react";

function App() {
  const { currentUser } = useSelector((store) => store.currentUser);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getUsers());
    dispatch(getQuestions());
  }, [dispatch]);

  if (currentUser === "") {
    return <Login />;
  }

  return (
    <main>
      <Navbar />
      <Routes>
        <Route path="*" element={<Error />} />
        <Route path="/" on exact element={<Home />}></Route>
        <Route path="/addpoll" exact element={<Addpoll />}></Route>
        <Route path="/leaderboard" exact element={<Leaderboard />}></Route>
        <Route path="/login" exact element={<Login />}></Route>

        <Route path="/questions/:id" element={<Question />}></Route>
      </Routes>
    </main>
  );
}
export default App;
