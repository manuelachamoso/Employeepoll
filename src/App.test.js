import React from "react";
import { render, screen } from "@testing-library/react";
import App from "./App";
import { Provider } from "react-redux";
import { store } from "./features/store";
import { MemoryRouter } from "react-router-dom";
import {setUser} from './reducers/userSlice'


describe("App", () => {
    it("will render the App component properly", () => {
        let view = render(
            <MemoryRouter>
                <Provider store={ store }>
                    <App />
                </Provider>
            </MemoryRouter>
      );
      expect(view).toMatchSnapshot();
    });
  
    it("will show home page when user is logged in correctly", () => {
      store.dispatch(setUser({ id: "zoshikanlu", password: "pass246" }));
      render(
        <MemoryRouter>
                <Provider store={ store }>
                    <App />
                </Provider>
            </MemoryRouter>
      );
      expect(screen.getByText(/Welcome/i)).toBeInTheDocument();
    });
  });