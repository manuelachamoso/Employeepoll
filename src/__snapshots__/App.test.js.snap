// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App will render the App component properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <div
          class="container-navbar"
        >
          <nav
            class="navbar"
          >
            <div
              class="nav-bar"
            >
              <button
                class="nav-button"
              >
                <a
                  href="/"
                >
                  Home
                </a>
              </button>
              <button
                class="nav-button"
              >
                <a
                  href="/add"
                >
                  Add Poll
                </a>
              </button>
              <button
                class="nav-button"
              >
                <a
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </button>
            </div>
            <div
              class="nav-bar-two"
            >
              <p />
            </div>
          </nav>
        </div>
        <div
          class="login"
        >
          <div
            class="profile"
          >
            <img
              alt=""
              src="https://cdn0.iconfinder.com/data/icons/unigrid-flat-human-vol-2/90/011_101_anonymous_anonym_hacker_vendetta_user_human_avatar-512.png"
            />
          </div>
          <h1
            class="header"
          >
            Employee Polls
          </h1>
          <h1
            class="login-title"
          >
            Login
          </h1>
          <form
            class="login-form"
          >
            Username:
            <input
              class="username"
              data-testid="username"
              id="username"
              name="username"
              required=""
              type="text"
            />
            Password:
            <input
              class="password"
              data-testid="password"
              id="password"
              name="password"
              required=""
              type="password"
            />
            <input
              class="button-submit"
              data-testid="submit"
              type="submit"
            />
            <div />
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <div
        class="container-navbar"
      >
        <nav
          class="navbar"
        >
          <div
            class="nav-bar"
          >
            <button
              class="nav-button"
            >
              <a
                href="/"
              >
                Home
              </a>
            </button>
            <button
              class="nav-button"
            >
              <a
                href="/add"
              >
                Add Poll
              </a>
            </button>
            <button
              class="nav-button"
            >
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </button>
          </div>
          <div
            class="nav-bar-two"
          >
            <p />
          </div>
        </nav>
      </div>
      <div
        class="login"
      >
        <div
          class="profile"
        >
          <img
            alt=""
            src="https://cdn0.iconfinder.com/data/icons/unigrid-flat-human-vol-2/90/011_101_anonymous_anonym_hacker_vendetta_user_human_avatar-512.png"
          />
        </div>
        <h1
          class="header"
        >
          Employee Polls
        </h1>
        <h1
          class="login-title"
        >
          Login
        </h1>
        <form
          class="login-form"
        >
          Username:
          <input
            class="username"
            data-testid="username"
            id="username"
            name="username"
            required=""
            type="text"
          />
          Password:
          <input
            class="password"
            data-testid="password"
            id="password"
            name="password"
            required=""
            type="password"
          />
          <input
            class="button-submit"
            data-testid="submit"
            type="submit"
          />
          <div />
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
